src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,211 @@
+  // The cone on either side of cone_new_apex is ok, so we use abs below.
+  ASSERT(acos(abs(cos_alpha)) < abs(asin(cos_theta)),
+         "The arguments passed into the CylindricalEndcap constructor "
+         "yield a noninvertible map.");
+
+  // Another necessary condition for invertibility is that proj_center
+  // cannot lie between sphere_one and cone_new_apex.
+  const double proj_radius_one = sqrt(square(center_one[0] - proj_center[0]) +
+                                      square(center_one[1] - proj_center[1]) +
+                                      square(center_one[2] - proj_center[2]));
+  ASSERT(proj_center[2] > cone_new_apex[2] or proj_center[2] < z_plane or
+             proj_radius_one < radius_one,
+         "The arguments passed into the CylindricalEndcap constructor "
+         "yield a noninvertible map.");
+
+  // Other sanity checks that may be relaxed if there is more logic
+  // added and more unit tests to test these cases.
+
+  ASSERT(proj_center[2] < z_plane,
+         "CylindricalEndcap: The map hasn't been tested for this "
+         "configuration. The map may still be invertible, but further "
+         "testing would be needed to ensure that jacobians are not "
+         "ill-conditioned.");
+
+  ASSERT(abs(cos_theta) <= 0.9,
+         "CylindricalEndcap: z_plane is too far from the center of sphere_one. "
+             << "cos_theta = " << cos_theta
+             << ". If |cos_theta| > 1 the map is singular.  If 0.9 < "
+                "|cos_theta| < 1 then the map is not singular, but the "
+                "jacobians are likely to be large and the map has not been "
+                "tested for these parameters.");
+  ASSERT(
+      abs(cos_theta) >= 0.1,
+      "CylindricalEndcap: z_plane is too close to the center of sphere_one. "
+          << "cos_theta = " << cos_theta
+          << ". The map is not singular, but the jacobians are likely to be "
+             "large "
+             "and the map has not been tested for this choice of parameters.");
+
+  const double dist_spheres = sqrt(square(center_one[0] - center_two[0]) +
+                                   square(center_one[1] - center_two[1]) +
+                                   square(center_one[2] - center_two[2]));
+  ASSERT(dist_spheres + radius_one < radius_two,
+         "CylindricalEndcap: The map has been tested only for the case when "
+         "sphere_one is contained inside sphere_two");
+
+  const double proj_radius_two = sqrt(square(center_two[0] - proj_center[0]) +
+                                      square(center_two[1] - proj_center[1]) +
+                                      square(center_two[2] - proj_center[2]));
+  ASSERT(proj_radius_two < radius_two,
+         "CylindricalEndcap: The map has been tested only for the case when "
+         "proj_center is contained inside sphere_two");
+
+  // Check if we are too close to singular.
+  ASSERT(acos(abs(cos_alpha)) < abs(0.95 * asin(cos_theta)),
```

Comment:
Where does 0.95 come from? Also, this renders the check on line 78 redundant, so why do both checks? If you have a good reason to do both, why not do them one right after the other?